### PR TITLE
ROX-16811: Create a custom RHEL updater

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Add a custom RHEL updater: It contains RH CVSS scores and handles unfixed Openshift vulns.
 
Diff between ClairCore and the custom RHEL updater without the customizations:

```diff
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: coalescer.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: coalescer_test.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: distributionscanner.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: distributionscanner_test.go
Only in /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel: doc.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: dockerfile
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: ecosystem.go
Only in /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/internal/common: charset.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/internal/common: updater.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/internal: containerapi
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: matcher.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: matcher_test.go
diff -bdur /home/jvdm/wt/github.com/quay/claircore/main/rhel/parser.go /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/parser.go
--- /home/jvdm/wt/github.com/quay/claircore/main/rhel/parser.go	2023-11-03 20:16:13.282361986 -0300
+++ /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/parser.go	2023-12-19 16:49:22.483207860 -0300
@@ -10,10 +10,9 @@
 	"github.com/quay/zlog"
 
 	"github.com/quay/claircore"
-	"github.com/quay/claircore/internal/xmlutil"
 	"github.com/quay/claircore/pkg/ovalutil"
-	"github.com/quay/claircore/rhel/internal/common"
 	"github.com/quay/claircore/toolkit/types/cpe"
+	"github.com/stackrox/rox/scanner/updater/rhel/internal/common"
 )
 
 // Parse implements [driver.Updater].
@@ -28,7 +27,7 @@
 	defer r.Close()
 	root := oval.Root{}
 	dec := xml.NewDecoder(r)
-	dec.CharsetReader = xmlutil.CharsetReader
+	dec.CharsetReader = common.CharsetReader
 	if err := dec.Decode(&root); err != nil {
 		return nil, fmt.Errorf("rhel: unable to decode OVAL document: %w", err)
 	}
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: repositoryscanner.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: repositoryscanner_test.go
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel: rhcc
diff -bdur /home/jvdm/wt/github.com/quay/claircore/main/rhel/rhel.go /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/rhel.go
--- /home/jvdm/wt/github.com/quay/claircore/main/rhel/rhel.go	2023-12-06 17:36:04.217752009 -0300
+++ /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/rhel.go	2023-12-19 16:49:22.483207860 -0300
@@ -22,6 +22,7 @@
 var (
 	_ driver.Updater      = (*Updater)(nil)
 	_ driver.Configurable = (*Updater)(nil)
+	repositoryKey                     = "rhel-cpe-repository"
 )
 
 // Updater fetches and parses RHEL-flavored OVAL databases.
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: com.redhat.rhsa-RHEL8.xml
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: layer-with-cpe.tar
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: layer-with-embedded-cs.tar
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: layer-with-no-cpe-info.tar
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: Red_Hat_Enterprise_Linux_5.xml
Only in /home/jvdm/wt/github.com/quay/claircore/main/rhel/testdata: releasefiles
diff -bdur /home/jvdm/wt/github.com/quay/claircore/main/rhel/updaterset.go /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/updaterset.go
--- /home/jvdm/wt/github.com/quay/claircore/main/rhel/updaterset.go	2023-10-12 03:58:11.221035023 -0300
+++ /home/jvdm/wt/github.com/stackrox/stackrox/rhscores/scanner/updater/rhel/updaterset.go	2023-12-19 16:49:22.535208026 -0300
@@ -12,7 +12,7 @@
 	"github.com/quay/zlog"
 
 	"github.com/quay/claircore/libvuln/driver"
-	"github.com/quay/claircore/rhel/internal/pulp"
+	"github.com/stackrox/rox/scanner/updater/rhel/internal/pulp"
 )
 
 // DefaultManifest is the url for the Red Hat OVAL pulp repository.
```
## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scanner/updater/export.go [hunk 1]
```diff
@@ -56,42 +57,48 @@ func Export(ctx context.Context, outputDir string) error {
...
+
+	// Manual CVEs.
 	manualSet, err := manual.UpdaterSet(ctx, nil)
 	if err != nil {
 		return err
 	}
-
 	outOfTree := append([]driver.Updater{}, manualSet.Updaters()...)
+	opts = append(opts, []updates.ManagerOption{
+		updates.WithOutOfTree(outOfTree)})
```
Comment (RTann) [inline scanner/updater/export.go:69]: nit: single line?

**Review by RTann — COMMENTED** [review]
Thanks for doing this. Can we clean up things we don't need like a bunch of the test-related files so we don't overbloat the repo? Also, can we add a `rhel/doc.go` which documents the purpose of this and why we did it. Similar to https://github.com/stackrox/claircore/pull/1/files

File: scanner/updater/rhel/doc.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// happen.
+//
+// The contents of this package are almost entirely copied from claircore[2],
+// except for:
+//
+//  1. The [Severity] field's contents are replaced with an urlencoded key/values
+//     containing the severity and additional CVSS base score/vector, and
+//
+//  2. A workaround to overcome the limitations of OVAL regarding Openshift
+//     unfixed vulns.
```
Comment (RTann) [inline scanner/updater/rhel/doc.go:20]: nit: unfixed OpenShift vulns

File: scanner/updater/rhel/parse_test.go [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+
+			if !cmp.Equal(cpes, test.expected) {
+				t.Fatal(cmp.Diff(cpes, test.expected))
+			}
+		})
+	}
+}
+
+// Here's a giant restructured struct for reference and tests.
+var ovalDef = oval.Definition{
```
Comment (RTann) [inline scanner/updater/rhel/parse_test.go:202]: this is unused and can be deleted to minimize the file size

File: scanner/updater/export.go [hunk 4]
```diff
@@ -56,42 +57,51 @@ func Export(ctx context.Context, outputDir string) error {
...
+	var opts [][]updates.ManagerOption
+
+	// Manual CVEs.
 	manualSet, err := manual.UpdaterSet(ctx, nil)
 	if err != nil {
 		return err
 	}
-
 	outOfTree := append([]driver.Updater{}, manualSet.Updaters()...)
+	opts = append(opts, []updates.ManagerOption{updates.WithOutOfTree(outOfTree)})
```
Comment (RTann) [inline scanner/updater/export.go:68]: are we able to just do 

```
opts = append(opts, []updates.ManagerOption{updates.WithOutOfTree(manualSet.Updaters())})
```

File: scanner/updater/export.go [hunk 5]
```diff
@@ -56,42 +57,51 @@ func Export(ctx context.Context, outputDir string) error {
...
+
+	// RHEL custom: Forked from ClairCore.
+	fac, err := rhel.NewFactory(ctx, rhel.DefaultManifest)
+	if err != nil {
+		return err
+	}
+	opts = append(opts, []updates.ManagerOption{
+		updates.WithFactories(map[string]driver.UpdaterSetFactory{"rhel-custom": fac})})
```
Comment (RTann) [inline scanner/updater/export.go:76]: I believe we'll need `updates.WithConfigs` as well so we can enable the unfixed vulns.

File: scanner/updater/rhel/parser.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+					Str("CVSS2", cve.Cvss2).
+					Msg("unexpected format")
+				continue
+			}
+			parsedScore, err := strconv.ParseFloat(score, 64)
+			if err != nil {
+				zlog.Warn(ctx).
+					Str("Vulnerability", def.Title).
+					Err(err).
+					Msg("parsing CVSS3")
```
Comment (RTann) [inline scanner/updater/rhel/parser.go:175]: CVSS2

File: scanner/updater/rhel/parser.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+				cvss3.vector = vector
+			}
+		}
+
+		if cve.Cvss2 != "" {
+			score, vector, found := strings.Cut(cve.Cvss2, "/")
+			if !found {
+				zlog.Warn(ctx).
+					Str("CVSS2", cve.Cvss2).
+					Msg("unexpected format")
```
Comment (RTann) [inline scanner/updater/rhel/parser.go:167]: can you add the vuln title to this log and the log we have for CVSS3, too?

**Comment by RTann** [comment]
It may also be worth looking at which files/tests we don't really care about because we didn't touch so we can minimize the size of this change (and the repo once this is merge)

**Comment by jvdm** [comment]
> It may also be worth looking at which files/tests we don't really care about because we didn't touch so we can minimize the size of this change (and the repo once this is merge)

It's fine leaving some garbage behind: It's easier to compare with ClairCore, and it will be deleted in the next release (or eventually).

File: scanner/updater/rhel/parser.go (hunk 7 above)
Comment (jvdm) [reply scanner/updater/rhel/parser.go:167]: 34502f8bc8

File: scanner/updater/rhel/parser.go (hunk 6 above)
Comment (jvdm) [reply scanner/updater/rhel/parser.go]: d2bb36ca34

File: scanner/updater/export.go (hunk 5 above)
Comment (jvdm) [reply scanner/updater/export.go:76]: [I thought it was ON by default](https://github.com/jvdm/quay-claircore-dev/blob/26b7d74e729c6e3850710b1f91934c4b883f92c9/rhel/rhel.go#L32).

File: scanner/updater/export.go (hunk 4 above)
Comment (jvdm) [reply scanner/updater/export.go:68]: I assumed there was a reason, and haven't checked, so I left it alone.

File: scanner/updater/rhel/parse_test.go (hunk 3 above)
Comment (jvdm) [reply scanner/updater/rhel/parse_test.go]: fc252c35ea

File: scanner/updater/rhel/doc.go (hunk 2 above)
Comment (jvdm) [reply scanner/updater/rhel/doc.go]: 97f033ca1b

File: scanner/updater/export.go (hunk 4 above)
Comment (RTann) [reply scanner/updater/export.go:68]: I don't think there really was a reason, but we can keep it as-is

File: scanner/updater/export.go (hunk 5 above)
Comment (RTann) [reply scanner/updater/export.go:76]: you're right i flipped the configuration around

**Review by RTann — APPROVED** [review]

**Pull request merged**